src/drivers/intel/cavs/mn.c

```diff
@@ -61,28 +85,49 @@ int mn_set_mclk(uint16_t mclk_id, uint32_t mclk_rate)
 	}
 
 	spin_lock(mn_lock);
-	mdivc = mn_reg_read(0x0);
 
-	/* Enable MCLK Divider */
-	mdivc |= 0x1;
+	mclk_sources[mclk_id] = MN_MCLK_SOURCE_NONE;
 
-	/* searching the smallest possible mclk source */
-	for (i = MAX_SSP_FREQ_INDEX; i >= 0; i--) {
-		if (mclk_rate > ssp_freq[i].freq)
-			break;
+	if (is_mclk_source_in_use()) {
+		clk_index = mclk_source_clock;
 
-		if (ssp_freq[i].freq % mclk_rate == 0)
-			clk_index = i;
-	}
+		if (ssp_freq[clk_index].freq % mclk_rate != 0) {
+			trace_mn_error("error: MCLK %d, no valid configuration for already selected source = %d",
+				       mclk_rate, mclk_source_clock);
+			ret = -EINVAL;
+			goto out;
+		}
+	} else {
```

Comment:
we can remove the else and reduce the indentation here because of the goto above

Reply:
oh no wait, sorry I misread it. ignore this comment